deploy-service/teletraanservice/src/main/java/com/pinterest/teletraan/worker/AgentJanitor.java

```diff
@@ -87,113 +93,60 @@ private void processStaleHosts(Collection<String> staleHostIds, boolean isMarked
     }
 
     private Long getInstanceLaunchGracePeriod(String clusterName) throws Exception {
-        Long launchGracePeriod = rodimusManager.getClusterInstanceLaunchGracePeriod(clusterName);
+        Long launchGracePeriod = (clusterName != null) ? rodimusManager.getClusterInstanceLaunchGracePeriod(clusterName) : null;
         return launchGracePeriod == null ? maxLaunchLatencyThreshold : launchGracePeriod * 1000;
     }
 
-    @Override
-    void processIndividualHosts() throws Exception {
-        long current_time = System.currentTimeMillis();
 
+    private void processLowWatermarkHosts() throws Exception {
+        long current_time = System.currentTimeMillis();
         // If host fails to ping for longer than min stale threshold,
         // either mark them as UNREACHABLE, or remove if confirmed with source of truth
         long minThreshold = current_time - minStaleHostThreshold;
-        List<HostBean> minStaleHosts = hostDAO.getStaleHosts(minThreshold);
+        List<HostAgentBean> minStaleHosts = hostAgentDAO.getStaleHosts(minThreshold);
         Set<String> minStaleHostIds = new HashSet<>();
-        for (HostBean host : minStaleHosts) {
-            minStaleHostIds.add(host.getHost_id());
+        for (HostAgentBean hostAgentBean : minStaleHosts) {
+            minStaleHostIds.add(hostAgentBean.getHost_id());
+        }
+        Collection<String> terminatedHosts = getTerminatedHostsFromSource(minStaleHostIds);
+        for (String removedId : terminatedHosts) {
+            removeStaleHost(removedId);
         }
-        processStaleHosts(minStaleHostIds, false);
+    }
 
+    private void processHighWatermarkHosts() throws Exception {
+        long current_time = System.currentTimeMillis();
         current_time = System.currentTimeMillis();
         long maxThreshold = current_time - Math.min(maxStaleHostThreshold, maxLaunchLatencyThreshold);
-        List<HostBean> maxStaleHosts = hostDAO.getStaleHosts(maxThreshold);
+        List<HostAgentBean> maxStaleHosts = hostAgentDAO.getStaleHosts(maxThreshold);
         Set<String> staleHostIds = new HashSet<>();
-        for (HostBean host : maxStaleHosts) {
+        
+        for (HostAgentBean hostAgentBean : maxStaleHosts) {
+            Long launchGracePeriod = getInstanceLaunchGracePeriod(hostAgentBean.getAuto_scaling_group());
+            HostBean host = hostDAO.getHostsByHostId(hostAgentBean.getHost_id()).get(0);
             if (host.getState() == HostState.PROVISIONED) {
-                if (current_time - host.getLast_update() >= maxLaunchLatencyThreshold) {
-                    staleHostIds.add(host.getHost_id());
+                if (current_time - hostAgentBean.getLast_update() >= launchGracePeriod) {
```

Comment:
Maybe you can merge the two ifs.
Also can extract the ifelse logic as well.
Not a requirement, just an option

Reply:
thought of not touching this part on the first go but I think it makes sense.

Reply:
diff updated.